include/envoy/request_info/request_info.h

```diff
@@ -71,63 +71,123 @@ class RequestInfo {
 
   /**
-   * @return the protocol of the request.
+   * Set the monotonic clock time when the last byte of the request was received.
    */
-  virtual const Optional<Http::Protocol>& protocol() const PURE;
+  virtual void lastDownstreamRxByteReceived(MonotonicTime) PURE;
 
   /**
-   * Set the request's protocol.
+   * @return the monotonic clock time when the first byte of the request was sent upstream. There
+   * may be a considerable delta between lastDownstreamByteReceived and this value due to filters.
    */
-  virtual void protocol(Http::Protocol protocol) PURE;
+  virtual const Optional<MonotonicTime>& firstUpstreamTxByteSent() const PURE;
 
   /**
-   * @return the response code.
+   * Set the monotonic clock time when the first byte of the request was sent upstream.
    */
-  virtual const Optional<uint32_t>& responseCode() const PURE;
+  virtual void firstUpstreamTxByteSent(MonotonicTime) PURE;
 
   /**
-   * @return the # of body bytes sent in the response.
+   * @return the monotonic clock time when the last bye of the request was sent upstream.
    */
-  virtual uint64_t bytesSent() const PURE;
+  virtual const Optional<MonotonicTime>& lastUpstreamTxByteSent() const PURE;
+
+  /**
+   * Set the monotonic clock time when the last bye of the request was sent upstream.
+   */
+  virtual void lastUpstreamTxByteSent(MonotonicTime) PURE;
 
   /**
-   * @return the microseconds duration of the first byte received to the last byte sent.
+   * @return the monotonic clock time when the first byte of the response is received from upstream.
    */
-  virtual std::chrono::microseconds duration() const PURE;
+  virtual const Optional<MonotonicTime>& firstUpstreamRxByteReceived() const PURE;
+
+  /**
+   * Set the monotonic clock time when the first byte of the response is received from upstream.
+   */
+  virtual void firstUpstreamRxByteReceived(MonotonicTime) PURE;
+
+  /**
+   * @return the monotonic clock time when the last byte of the response is received from upstream.
+   */
+  virtual const Optional<MonotonicTime>& lastUpstreamRxByteReceived() const PURE;
+
+  /**
+   * Set the monotonic clock time when the last byte of the response is received from upstream.
+   */
+  virtual void lastUpstreamRxByteReceived(MonotonicTime) PURE;
+
+  /**
+   * @return the monotonic clock time when the first byte of the response is sent downstream. There
+   * may be a considerable delta between lastUpstreamByteReceived and this value due to filters.
+   */
+  virtual const Optional<MonotonicTime>& firstDownstreamTxByteSent() const PURE;
+
+  /**
+   * Set the monotonic clock time when the first byte of the response is sent downstream.
+   */
+  virtual void firstDownstreamTxByteSent(MonotonicTime) PURE;
+
+  /**
+   * @return the monotonic clock time when the last byte of the response is sent downstream.
+   */
+  virtual const Optional<MonotonicTime>& lastDownstreamTxByteSent() const PURE;
```

Comment:
Most of the times in request_info have been optionals since I added access logging to tcp_proxy, because they don't make sense in all contexts, so this lets the interface return "no value set" in those cases.

Reply:
Also, from discussion with @mattklein123, if we don't actually make an upstream request (eg, an inline static response), we don't want upstream-specific timings.